testutils/testcluster: TestStopServer failed

testutils/testcluster.TestStopServer [failed](https://mesolite.cluster.engflow.com/invocations/default/8e81072d-4251-45da-bd30-cef841873756?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cvdGVzdHV0aWxzL3Rlc3RjbHVzdGVyOnRlc3RjbHVzdGVyX3Rlc3Q=) on release-26.1 @ [5122355c4cddec33691f2bedb42feb36b7785070](https://github.com/cockroachdb/cockroach/commits/5122355c4cddec33691f2bedb42feb36b7785070):


```
=== RUN   TestStopServer
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestStopServer2863907427
    test_log_scope.go:82: use -show-logs to present logs inline
    testcluster_test.go:255: status: 500 Internal Server Error, content-type: application/json, body: {
          "error": "An internal server error has occurred. Please check your CockroachDB logs for more details.",
          "code": 13,
          "message": "An internal server error has occurred. Please check your CockroachDB logs for more details.",
          "details": [
          ]
        }, error: <nil>
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestStopServer2863907427
--- FAIL: TestStopServer (29.02s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #163847 testutils/testcluster: TestStopServer failed [C-test-failure O-robot T-db-server branch-master release-blocker]
</p>
</details>
/cc @cockroachdb/test-eng
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStopServer.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-60534

--- a/pkg/server/status/recorder.go
+++ b/pkg/server/status/recorder.go
@@ -892,7 +892,11 @@ func extractValue(name string, mtr interface{}, fn func(string, float64)) error 
 		// NB: this branch is intentionally at the bottom since all metrics implement it.
 		m := mtr.ToPrometheusMetric()
 		if m.Gauge != nil {
-			fn(name, *m.Gauge.Value)
+			v := *m.Gauge.Value
+			if math.IsInf(v, 0) || math.IsNaN(v) {
+				v = 0
+			}
+			fn(name, v)
 		} else if m.Counter != nil {
 			fn(name, *m.Counter.Value)
 		}

--- a/pkg/server/status/runtime.go
+++ b/pkg/server/status/runtime.go
@@ -1158,7 +1158,7 @@ func (rsr *RuntimeStatSampler) SampleEnvironment(ctx context.Context, cs *CGoMem
 	hostNiceTime := int64(cpuUsage.Nice * 1.e9)
 
 	var procUrate, procSrate, hostUrate, hostSrate, hostIrqrate, hostSoftIrqrate, hostNiceRate float64
-	if rsr.last.now != 0 { // We cannot compute these rates on the first iteration.
+	if rsr.last.now != 0 && dur > 0 { // We cannot compute these rates on the first iteration or if no time elapsed.
 		procUrate = float64(procUtime-rsr.last.procUtime) / dur
 		procSrate = float64(procStime-rsr.last.procStime) / dur
 		hostUrate = float64(hostUtime-rsr.last.hostUtime) / dur
@@ -1180,14 +1180,20 @@ func (rsr *RuntimeStatSampler) SampleEnvironment(ctx context.Context, cs *CGoMem
 	gcStopTotalNs := int64(gcStopTotal * 1.e9)
 	nonGcStopTotalNs := int64(nonGcStopTotal * 1.e9)
 	gcCount := rsr.goRuntimeSampler.uint64(runtimeMetricGCCount)
-	gcPauseRatio := float64(gcPauseTotalNs-rsr.last.gcPauseTime) / dur
+	var gcPauseRatio float64
+	if dur > 0 {
+		gcPauseRatio = float64(gcPauseTotalNs-rsr.last.gcPauseTime) / dur
+	}
 	runnableSum := goschedstats.CumulativeNormalizedRunnableGoroutines()
 	gcAssistSeconds := rsr.goRuntimeSampler.float64(runtimeMetricGCAssist)
 	gcAssistNS := int64(gcAssistSeconds * 1e9)
 	// The number of runnable goroutines per CPU is a count, but it can vary
 	// quickly. We don't just want to get a current snapshot of it, we want the
 	// average value since the last sampling.
-	runnableAvg := (runnableSum - rsr.last.runnableSum) * 1e9 / dur
+	var runnableAvg float64
+	if dur > 0 {
+		runnableAvg = (runnableSum - rsr.last.runnableSum) * 1e9 / dur
+	}
 	rsr.last.now = now
 	rsr.last.procUtime = procUtime
 	rsr.last.procStime = procStime
@@ -1201,7 +1207,10 @@ func (rsr *RuntimeStatSampler) SampleEnvironment(ctx context.Context, cs *CGoMem
 
 	// Log summary of statistics to console.
 	osStackBytes := rsr.goRuntimeSampler.uint64(runtimeMetricMemStackOSBytes)
-	cgoRate := float64((numCgoCall-rsr.last.cgoCall)*int64(time.Second)) / dur
+	var cgoRate float64
+	if dur > 0 {
+		cgoRate = float64((numCgoCall-rsr.last.cgoCall)*int64(time.Second)) / dur
+	}
 	goAlloc := rsr.goRuntimeSampler.uint64(runtimeMetricHeapAlloc)
 	goTotal := rsr.goRuntimeSampler.uint64(runtimeMetricGoTotal) -
 		rsr.goRuntimeSampler.uint64(runtimeMetricHeapReleasedBytes)

--- a/pkg/util/metric/metric.go
+++ b/pkg/util/metric/metric.go
@@ -1182,7 +1182,11 @@ func (g *GaugeFloat64) Inspect(f func(interface{})) { f(g) }
 
 // MarshalJSON marshals to JSON.
 func (g *GaugeFloat64) MarshalJSON() ([]byte, error) {
-	return json.Marshal(g.Value())
+	v := g.Value()
+	if math.IsInf(v, 0) || math.IsNaN(v) {
+		v = 0
+	}
+	return json.Marshal(v)
 }
 
 // ToPrometheusMetric returns a filled-in prometheus metric of the right type.

--- a/pkg/util/metric/metric_test.go
+++ b/pkg/util/metric/metric_test.go
@@ -119,6 +119,22 @@ func TestGaugeFloat64(t *testing.T) {
 	}
 }
 
+func TestGaugeFloat64MarshalInfNaN(t *testing.T) {
+	g := NewGaugeFloat64(emptyMetadata)
+
+	g.Update(math.Inf(1))
+	testMarshal(t, g, "0")
+
+	g.Update(math.Inf(-1))
+	testMarshal(t, g, "0")
+
+	g.Update(math.NaN())
+	testMarshal(t, g, "0")
+
+	g.Update(42.5)
+	testMarshal(t, g, "42.5")
+}
+
 func TestCounter(t *testing.T) {
 	c := NewCounter(emptyMetadata)
 	c.Inc(90)

